invenio_app_ils/permissions.py

```diff
@@ -79,6 +97,53 @@ def files_permission(obj, action=None):
     return backoffice_permission()
 
 
+def is_loan_owner_permission(patron_pid):
+    """Return permission to allow loan owner, librarians and admins."""
+    return Permission(UserNeed(int(patron_pid)), backoffice_access_action)
+
+
+def loan_extend_circulation_permission(loan):
+    """Return permission to allow only owner and librarians to extend loan."""
+    if not current_user or not current_user.id:
+        abort(401)
+
+    if current_user.id == int(loan["patron_pid"]):
+        validate_automatic_loan_extension(loan)
+
+    return is_loan_owner_permission(loan["patron_pid"])
+
+
+def validate_automatic_loan_extension(loan):
+    """Validate if the loan can be extended automatically."""
+    # NOTE: Validate max extensions
+    if not (
+        loan["extension_count"] <=
+        circulation_default_extension_max_count(loan)
+    ):
+        raise InvalidLoanExtendError(
+            "You have reached the max number of extensions for this loan!"
+        )
+
+    if not validate_overbook(loan):
+        raise InvalidLoanExtendError("This document is overbooked!")
+
+
+def validate_overbook(loan):
+    """Validate if the document of the loan is overbooked."""
+    doc_pid = loan["document_pid"]
+    loan_search = current_circulation.loan_search_cls()
+    pending_loans = loan_search.get_pending_loans_by_doc_pid(doc_pid).count()
+    active_loans = loan_search.get_active_loans_by_doc_pid(doc_pid).count()
+
+    item_search = current_app_ils.item_search_cls()
+    items = item_search.search_by_document_pid(doc_pid).count()
+    unavailable_items = item_search.get_unavailable_items_by_document_pid(
+        doc_pid).count()
+
+    remain_items = items - active_loans - unavailable_items
+    return pending_loans > remain_items
```

Comment:
why not simply re-using the resolver?
```
doc_pid = loan["document_pid"]
Document = current_app_ils.document_record_cls()
doc = Document.get_record_by_pid(doc_pid)
doc.replace_refs()  # this is needed only if refs are not computed
return doc and doc["circulation"]["overbooked"]
```
WDYT?

Reply:
Anything wrong with `loan = loan.replace_refs()`?